### DISCUSSION TITLE
Using an async fn in set_on_submit

### DISCUSSION THREAD

**Discussion by nhtyy** [gyscos/cursive#692] [Q&A]
trying to get this to work any suggestions?

Code that calls the callback
```
    pub fn set_on_submit<F, R, V: ?Sized>(&mut self, cb: F)
    where
        F: 'static + Fn(&mut Cursive, &V) -> R,
        T: Borrow<V>,
    {
        self.on_submit = Some(Rc::new(move |s, t| {
            cb(s, t.borrow());
        }));
    }
```

example callback using tokio as runtime btw
```
async fn handle_function_call<F>(siv: &mut Cursive, func: &F) {
   // work work work
  ()
}
```

example error (note the expected and found match??)
```"mismatched types expected trait `for<'r, 's> <for<'r, 's> fn(&'r mut Cursive, &'s AryaFunction) -> impl for<'r, 's> Future<Output = ()> {handle_function_call::<AryaFunction>} as FnOnce<(&'r mut Cursive, &'s AryaFunction)>>`\n   found trait `for<'r, 's> <for<'r, 's> fn(&'r mut Cursive, &'s AryaFunction) -> impl for<'r, 's> Future<Output = ()> {handle_function_call::<AryaFunction>} as FnOnce<(&'r mut Cursive, &'s AryaFunction)>>`",```

adding lifetime bounds
```
async fn handle_function_call<'r, 's, F>(siv: &'r mut Cursive, func: &'s F) {
   // work work work
  ()
}
```

yields same error

**Comment by gyscos** [comment]
Hi, and thanks for the question!

At the moment cursive is not async aware.
This means that most views use non-async functions for callbacks.
What you can do instead is to use a regular closure in which you spawn an async task (and possibly wait for completion if you need a result there, but ideally not).

Now one issue is if you need to borrow `&mut Cursive`. It'd be easier if you didn't.
You should instead:
* In the (non async) callback closure, first get anything you need from the &mut Cursive` (state from other views, maybe clone its `cb_sink`, ...).
* Spawn an async task to run async processing (maybe network connections). You would _move_ the stuff you need there. This async task would not take any reference to `Cursive` and ideally would be `'static`.
* Return from the callback, with the async task still running. Optionally indicate that a background operation is running (show a popup for instance, or lock the UI to avoid receiving another submit callback before the current task completes).
* From the async task, use the cloned `cb_sink` to report the results to the UI.

Overall it's very similar to using a background thread, but using an async task instead.

**Reply by nhtyy** [reply]
Hey thanks for the edit, appreciate your time!
